homeassistant/helpers/script.py

```diff
@@ -130,39 +579,50 @@ def __init__(
         sequence: Sequence[Dict[str, Any]],
         name: Optional[str] = None,
         change_listener: Optional[Callable[..., Any]] = None,
+        if_running: Optional[str] = None,
+        run_mode: Optional[str] = None,
+        logger: Optional[logging.Logger] = None,
+        log_exceptions: bool = True,
     ) -> None:
         """Initialize the script."""
-        self.hass = hass
+        self._logger = logger or logging.getLogger(__name__)
+        self._hass = hass
         self.sequence = sequence
         template.attach(hass, self.sequence)
         self.name = name
         self._change_listener = change_listener
-        self._cur = -1
-        self._exception_step: Optional[int] = None
         self.last_action = None
         self.last_triggered: Optional[datetime] = None
         self.can_cancel = any(
             CONF_DELAY in action or CONF_WAIT_TEMPLATE in action
             for action in self.sequence
         )
-        self._async_listener: List[CALLBACK_TYPE] = []
+        if not if_running and not run_mode:
+            self._if_running = IF_RUNNING_PARALLEL
+            self._run_mode = RUN_MODE_LEGACY
+        elif if_running and run_mode == RUN_MODE_LEGACY:
+            self._raise('Cannot use if_running if run_mode is "legacy"')
+        else:
+            self._if_running = if_running or IF_RUNNING_CHOICES[0]
+            self._run_mode = run_mode or RUN_MODE_CHOICES[0]
+        self._runs: List[_ScriptRunBase] = []
+        self._log_exceptions = log_exceptions
         self._config_cache: Dict[Set[Tuple], Callable[..., bool]] = {}
-        self._actions = {
-            ACTION_DELAY: self._async_delay,
-            ACTION_WAIT_TEMPLATE: self._async_wait_template,
-            ACTION_CHECK_CONDITION: self._async_check_condition,
-            ACTION_FIRE_EVENT: self._async_fire_event,
-            ACTION_CALL_SERVICE: self._async_call_service,
-            ACTION_DEVICE_AUTOMATION: self._async_device_automation,
-            ACTION_ACTIVATE_SCENE: self._async_activate_scene,
-        }
         self._referenced_entities: Optional[Set[str]] = None
         self._referenced_devices: Optional[Set[str]] = None
 
+    def _changed(self):
+        if self._change_listener:
+            self._hass.async_add_job(self._change_listener)
+
+    def set_logger(self, logger):
+        """Set logger used by script helper."""
+        self._logger = logger
+
     @property
     def is_running(self) -> bool:
         """Return true if script is on."""
-        return self._cur != -1
+        return len(self._runs) > 0
```

Comment:
Slightly faster.

```suggestion
        return bool(self._runs)
```

Reply:
I had thought of doing that but decided using `len(x) > 0` was clearer. Also, I just tried both ways using `timeit` and, believe it or not, `len(x) > 0` is actually faster than `bool(x)`.